spec/lib/heartcheck/controllers/environment_spec.rb

```diff
@@ -0,0 +1,74 @@
+module Heartcheck
+  module Controllers
+    describe Environment do
+      subject(:controller) { described_class.new }
+      describe '#index' do
+        subject(:index) { MultiJson.load(controller.index, symbolize_keys: true) }
+
+        before do
+          stub_const('RUBY_VERSION', '2.4.0')
+        end
+
+        system_info_example = { :example_key => 'example_value' }
+
+        context "given Rails is used" do
+          before do
+            stub_const('Rails::VERSION::STRING', '5.0.0')
+          end
+
+          it 'gets info from the right constants and functions' do
+            allow(Sys::Uname).to receive_message_chain(:uname, :each_pair, :to_a).and_return([[:example_key, "example_value"]])
```

Comment:
To avoid this stub, what do you think to test the real result?

As a suggestion, it would be like this:
```ruby
expect(index.keys).to match_array([:system_info, :ruby_version, :rails_version])
expect(index[:system_info].keys).to match_array([:sysname, :nodename, :release, :version, :machine, :domainname])
```

Reply:
Yes, but I did test the real result in another test, starting on line 40:
```
          it 'gets the info in the expected format' do
            expect(system_info).to be_a(Hash)
            expect(system_info).to include(
              :sysname,
              :nodename,
              :release,
              :version,
              :machine,
              :domainname
            )
            expect(system_info[:sysname]).to be_a(String)
            expect(system_info[:nodename]).to be_a(String)
            expect(system_info[:release]).to be_a(String)
            expect(system_info[:version]).to be_a(String)
            expect(system_info[:machine]).to be_a(String)
            expect(system_info[:domainname]).to be_a(String)
            expect(ruby_version).to be_a(String)
            expect(rails_version).to be_a(String)
          end
```
Since the values of the constants RUBY_VERSION, Rails::VERSION::STRING and the return of the function Sys::Uname.uname are nondeterministic values, my idea was to "mock" these values just to make sure that at least the index method is getting them from the right place. What do you think? I am kind of new in implementing tests, so I appreciate some feedback =D

Reply:
humm... the test in line 40 is exactly what I suggested, so forget it!

But if you want to test `getting them from the right place` maybe you should add an expectation for it.

What about like this:
```ruby
it 'gets info from the right constants and functions' do
  allow(Sys::Uname).to receive_message_chain(:uname).and_return(system_info_example)

  is_expected.to include(
    :system_info => system_info_example,
    :ruby_version => '2.4.0',
    :rails_version => '5.0.0'
  )

  expect(Sys::Uname).to have_received(:uname)
end
```

Reply:
Good idea. But the `:uname` method returns a Struct object, that is why I did this message chain with `receive_message_chain(:uname, :each_pair, :to_a)`. What about like this?
```ruby
it 'gets info from the right constants and functions' do
  allow(Sys::Uname).to receive_message_chain(:uname, :each_pair, :to_a).and_return([[:example_key, "example_value"]])
  is_expected.to include(
    :system_info => system_info_example,
    :ruby_version => '2.4.0',
    :rails_version => '5.0.0'
  )
  expect(Sys::Uname).to have_received(:uname)
end
```

Reply:
I think the more you add these messages in `chain`, the more your spec will be coupled to the implementation, so that's why I'm trying to avoid it.

Doing the way I suggested, if someone changes the implementation to use only `Sys::Uname.uname`, another test will brake (I tested it)

wdyt?